core/src/execution/datafusion/expressions/cast.rs

```diff
@@ -635,6 +642,38 @@ impl Cast {
         Ok(cast_array)
     }
 
+    fn cast_string_to_date(
+        array: &ArrayRef,
+        to_type: &DataType,
+        eval_mode: EvalMode,
+    ) -> CometResult<ArrayRef> {
+        let string_array = array
+            .as_any()
+            .downcast_ref::<GenericStringArray<i32>>()
+            .expect("Expected a string array");
+
+        let cast_array: ArrayRef = match to_type {
+            DataType::Date32 => {
+                let len = string_array.len();
+                let mut cast_array = PrimitiveArray::<Date32Type>::builder(len);
+                for i in 0..len {
+                    if string_array.is_null(i) {
+                        cast_array.append_null()
+                    } else if let Ok(Some(cast_value)) =
+                        date_parser(string_array.value(i), eval_mode)
```

Comment:
@vidyasankarv This is the fix you need to make your current test pass.

The problem was that we were ignoring any error here when runnin in ANSI mode.

```suggestion
                    } else if let Some(cast_value) = date_parser(string_array.value(i), eval_mode)?
```

Reply:
oh sorry my mistake missed checking here.
handling all cases of return value of date_parser using matching clause now. 
thank you @andygrove